packages/grid/_modules_/grid/hooks/features/rows/useRows.ts

```diff
@@ -51,19 +61,23 @@ export const useRows = (rows: RowsProp, apiRef: ApiRef): void => {
 
   const internalRowsState = React.useRef<InternalRowsState>(gridState.rows);
 
+  const getRowId = React.useCallback(
+    (row: RowData) => {
+      return apiRef.current.getRowId(row);
+    },
+    [apiRef],
+  );
```

Comment:
The `apiRef` never changes during amounting lifecycle, the `getRowId` can. The logic is not correct:

```suggestion
  const getRowId = apiRef.current.getRowId(row);
```

Reply:
No because it is undefined on the first run as we have to run the hooks...

Reply:
otherwise I'll have to check for null and it's just better to put it in the react cycle using the useCallback I think

Reply:
> No because it is undefined on the first run as we have to run the hooks...

Ok, then the root problem is worse. We should **respect effect dependencies** as much as possible. When we update a prop, it should rerun all the effects. For instance, in https://codesandbox.io/s/material-demo-forked-7bg1k?file=/demo.js the fact that the warning for missing id isn't raising is a red flag.

Reply:
> > No because it is undefined on the first run as we have to run the hooks...
> 
> Ok, then the root problem is worse. We should **respect effect dependencies** as much as possible. The useCallback swallows it. When we update a prop, it should rerun all the effects. For instance, in https://codesandbox.io/s/material-demo-forked-7bg1k?file=/demo.js the fact that the warning for "missing id" isn't raising is a signal that the architecture is not correct.

I don't share the same opinion. I do not think the problem is 'worse', it's exactly the same root issue mentioned in the api provider POC PR.

Reply:
> I do not think the problem is 'worse'

You are right, I was referring to my understanding of the situation. It's in absolute, the same problem.